docs/dev-guide/guide.md

```diff
@@ -471,43 +471,89 @@ FOR NO KEY UPDATE;
+      requestedRole,
+      authzData,
+    }) {
+      assertHasRole(authzData, requestedRole, [
+        // The allowable roles that can call this function
+        'Student',
+        'Student Data Viewer',
+        'Student Data Editor',
+      ]);
+      const enrollment = await queryRow(sql.select_enrollment_by_id, { id }, EnrollmentSchema);
+      assertEnrollmentInCourseInstance(enrollment, courseInstance);
+      if (requestedRole === 'Student') {
+        assertEnrollmentBelongsToUser(enrollment, authzData);
+      }
+      return enrollment;
+    }
+    ```
+
+This is good, because in order to perform an update, you need to pass in a full row object, and a request body won't have enough information for this. Model functions that fetch rows require the caller to pass in the needed information to perform the correct authorization checks. For example, in the above example, the `selectEnrollment` function requires the caller to pass in the `courseInstance` and `authzData` parameters, so it can assert that the enrollment belongs to the user, and is in the correct course instance. It will throw an error if the caller is not authorized to access the enrollment (or null if it was `selectOptionalEnrollment`). This also forced the caller to prove access to the course instance in order to read the enrollment record.
+
+Once you have a full row object, you have asserted that the caller is authorized to _read_ the record. You will also need to assert that the caller is authorized to _write_ the record.
 
 ```typescript
-const enrollment = await dangerouslySelectEnrollment({ id: enrollment_id });
+await updateEnrollmentStatus({
+  enrollment: myEnrollment,
+  status: 'joined',
+  // What role are we trying to authorize as?
+  requiredRoleOptions: 'Student',
+  // Information to prove we are authorized to write the record
+  authzData: res.locals.authz_data,
+});
 ```
 
-There also may exist versions of these functions that don't require the caller to pass in the `course_instance` and `authLevel` parameters, like `dangerouslySelectEnrollment`, but these should be avoided if possible.
+In this example, instructors are typically allowed to read the enrollment record for students, but for certain actions, like joining a course instance, they need to be authorized as a student. The model function would note that the `requiredRoleOptions` parameter is `'Student'`, but the current user is an instructor, so it would throw an error.
+
+In some cases, you may not have access to `authzData`, e.g. if are pulling data from a queue, or deep in internal code. In this case, you can use the `dangerousFullAuthzForTesting` function to build a dummy `authzData` object that allows you to perform the action as the system. NOTE: We need to revisit this concept as we build out this pattern.
+
+```typescript
+const authzData = updateEnrollmentStatus({
+  enrollment: myEnrollment,
+  status: 'joined',
+  requiredRoleOptions: ['Student Data Viewer', 'Student Data Editor'],
```

Comment:
Is this meant to be either a string or array of strings? `requiredRoleOptions` looks weird for the single string case, but could make sense if you're expecting an array of multiple things.

That said: I'm not actually understanding why this is an array at all. "editor" implies "viewer", so both aren't needed. And if this is serving the same purpose as in `selectCourseInstance(...)` or `selectEnrollment(...)`, it especially feels like this should be a single value.

Given that this didn't change in 4392aa5, I'm assuming this name+shape is intentional? Idk, I sat here looking at this for quite a while and something is not clicking, I'm still confused. Can we discuss further?